Mage/src/main/java/mage/abilities/mana/ManaOptions.java

```diff
@@ -562,48 +555,13 @@ public boolean removeEqualMana(Mana manaToRemove) {
         return result;
     }
 
-    public static void addManaCombination(Mana mana, Mana existingMana, List<Mana> payCombinations, List<String> payCombinationsStrings) {
+    public static void addManaCombination(Mana mana, Mana existingMana, Set<Mana> payCombinations, Set<String> payCombinationsStrings) {
         Mana newMana = existingMana.copy();
         newMana.add(mana);
         payCombinations.add(newMana);
         payCombinationsStrings.add(newMana.toString());
     }
 
-    public void removeDuplicated() {
-        Set<String> list = new HashSet<>();
-
-        for (int i = this.size() - 1; i >= 0; i--) {
-            String s;
-            if (this.get(i) instanceof ConditionalMana) {
-                s = this.get(i).toString() + ((ConditionalMana) this.get(i)).getConditionString();
-            } else {
-                s = this.get(i).toString();
-            }
-            if (s.isEmpty()) {
-                this.remove(i);
-            } else if (list.contains(s)) {
-                // remove duplicated
-                this.remove(i);
-            } else {
-                list.add(s);
-            }
-        }
-
-        // Remove fully included variations
-        // TODO: research too many manas and freeze (put 1 card to slow down, put 3 cards to freeze here)
-        //  battlefield:Human:Cascading Cataracts:1
-        for (int i = this.size() - 1; i >= 0; i--) {
-            for (int ii = 0; ii < i; ii++) {
-                Mana moreValuable = Mana.getMoreValuableMana(this.get(i), this.get(ii));
-                if (moreValuable != null) {
-                    this.get(ii).setToMana(moreValuable);
-                    this.remove(i);
-                    break;
-                }
-            }
-        }
-    }
-
```

Comment:
This portion of the code did more than remove duplicates. It also removes the variations that are already included. E.g. if {B}{U} and {B} are in `this`, then {B} will be removed.

Is this a performance optimization or a requirement for how other parts of the game work?

Reply:
This scales part of the algorithm scales *ingredibly* poorly, `O(n^2)`.

Does anyone know if this is an actual requirement for other parts of the code? As in, is the code assuming that if ManaOptions has {B}{U} then it will *not* have {B} or {U}. Or the reverse, if it has {B} then {B}{U} is not available?

If it's an optimization, I am going to disable it since I'm not convinced it's worth the computation cost.

The tests (other than those which compare the size of manaOptions to a hardcoded value) all seem to pass. I will try playing some games against the AI with it disabled and see if I find anything broken.

Reply:
Looking through code, it looks like an optimization. I am removing it since it scales very poorly, and the performance penalty of not doing it appears to be less than the performance penalty of keeping it in.